packages/backend/src/database/migrations/20230116111752_add_pinned_list_table.ts

```diff
@@ -0,0 +1,73 @@
+import { Knex } from 'knex';
+
+const PinnedListTableName = 'pinned_list';
+const PinnedItemsTableName = 'pinned_items';
+
+export async function up(knex: Knex): Promise<void> {
+    await knex.schema.alterTable('projects', (table) => {
+        table.unique(['project_uuid']);
+    });
+    await knex.schema.alterTable('saved_queries', (table) => {
+        table.unique(['saved_query_uuid']);
+    });
+    await knex.schema.alterTable('dashboards', (table) => {
+        table.unique(['dashboard_uuid']);
+    });
+    if (!(await knex.schema.hasTable(PinnedListTableName))) {
+        await knex.schema.createTable(PinnedListTableName, (table) => {
+            table
+                .uuid('pinned_list_uuid')
+                .primary()
+                .notNullable()
+                .defaultTo(knex.raw('uuid_generate_v4()'));
+            table
+                .uuid('project_uuid')
+                .nullable()
+                .references('project_uuid')
+                .inTable('projects')
+                .onDelete('CASCADE');
+        });
+    }
+    if (!(await knex.schema.hasTable(PinnedItemsTableName))) {
+        await knex.schema.createTable(PinnedItemsTableName, (table) => {
+            table
+                .uuid('pinned_item_uuid')
+                .primary()
+                .notNullable()
+                .defaultTo(knex.raw('uuid_generate_v4()'));
+            table
+                .uuid('pinned_list_uuid')
+                .nullable()
+                .references('pinned_list_uuid')
+                .inTable(PinnedListTableName)
+                .onDelete('CASCADE');
+            table.string('pinned_item_type').notNullable();
+            table
+                .uuid('saved_chart_uuid')
```

Comment:
Can you Pin the same chart / dashboard multiple times ?

Reply:
Not to the same list. How can we add that restriction when the chart uuid might be null ?  cc: @owlas

Reply:
In this table structure it's pretty difficult because you have these mutually exclusive columns. It would be better to not have those two columns in one table. There are two alternatives:

- Create two tables `pinned_charts` and `pinned_dashboards` rather than trying to combine them into one.
- OR first create an `items` table and place a foreign key to an `item` on each of the `saved_charts` and `dashboards` tables. Then you can have a single foreign key on `pinned_items` to an `item` entry

The first one is the simplest. The second one is a bit of a refactor and might make more sense if we want to do things like list items, search items, etc.

Reply:
Someone previously mentioned that this is like a special space. So why not adding a flag to spaces  to make it a Pinning list ? It even makes sense if we include a different list in the future, like `starred` or `blacklisted` ...